### ISSUE TITLE
`explicit_iter_loop` false positive on `&Box<[_]>` with MSRV < 1.80

### ISSUE THREAD

**Issue by oxalica** [rust-lang/rust-clippy#13184]
### Summary

`for _ in box_slice.iter()` generates a `explicit_iter_loop` warning suggesting `for _ in &box_slice`, even with MSRV < 1.80. But the suggestion only compiles since 1.80 with the introduction of [`impl IntoIterator for &'_ Box<[_]>`](https://doc.rust-lang.org/1.80.0/std/iter/trait.IntoIterator.html#impl-IntoIterator-for-%26Box%3C%5BI%5D,+A%3E)

`box_slice.iter_mut()` also reproduces.

### Lint Name

explicit_iter_loop

### Reproducer

I tried this code:

```rust
#![warn(clippy::explicit_iter_loop)]

pub struct S {
    list: Box<[u32]>,
}

pub fn work(s: &S) {
    for m in s.list.iter() {
        dbg!(m);
    }
}
```

With this Cargo.toml:
```toml
[package]
name = "testt"
version = "0.1.0"
edition = "2021"
rust-version = "1.76"
```

Run `cargo clippy`, and I saw this happen:

```
warning: it is more concise to loop over references to containers instead of using explicit iteration methods
 --> src/main.rs:8:14
  |
8 |     for m in s.list.iter() {
  |              ^^^^^^^^^^^^^ help: to write this more concisely, try: `&s.list`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#explicit_iter_loop
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::explicit_iter_loop)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I expected to see this happen:

<no warning>

### Version

```text
rustc 1.80.0 (051478957 2024-07-21)
binary: rustc
commit-hash: 051478957371ee0084a7c0913941d2a8c4757bb9
commit-date: 2024-07-21
host: x86_64-unknown-linux-gnu
release: 1.80.0
LLVM version: 18.1.7
```
